phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexUpgradeTool.java

```diff
@@ -0,0 +1,550 @@
+            }
+            LOGGER.info("list of tables/indexes passed: " + inputTables);
+
+            tablesAndIndexes = extractTablesAndIndexes(inputTables);
+
+            if (operation.equalsIgnoreCase(UPGRADE_OP)) {
+                upgrade = true;
+            } else if (operation.equalsIgnoreCase(ROLLBACK_OP)) {
+                upgrade = false;
+            } else {
+                throw new IllegalStateException("Invalid option provided for "
+                        + OPERATION_OPTION.getOpt() + " expected values: {upgrade, rollback}");
+            }
+            if (dryRun) {
+                LOGGER.info("This is the beginning of the tool with dry run.");
+            }
+        } catch (IOException e) {
+            LOGGER.severe("Something went wrong "+e);
+        }
+    }
+
+    @VisibleForTesting
+    public void executeTool() {
+        try (Connection conn = DriverManager.getConnection(connectionURL, new Properties())) {
+            ConnectionQueryServices queryServices = conn.unwrap(PhoenixConnection.class)
+                    .getQueryServices();
+            Configuration conf = queryServices.getConfiguration();
+            executeTool(conn, queryServices, conf);
+        } catch (SQLException e) {
+            LOGGER.severe("Something went wrong in executing tool "+ e);
+        }
+    }
+
+    private void executeTool(Connection conn, ConnectionQueryServices queryServices,
+            Configuration conf) {
+
+        LOGGER.info("Executing "+ operation);
+
+        for (Map.Entry<String, ArrayList<String>> entry: tablesAndIndexes.entrySet()) {
+            String dataTableFullName = entry.getKey();
+            ArrayList<String> indexes = entry.getValue();
+
+            try (Admin admin = queryServices.getAdmin()) {
+
+                PTable dataTable = PhoenixRuntime.getTableNoCache(conn, dataTableFullName);
+                LOGGER.fine("Executing "+operation+" for " + dataTableFullName);
+
+                boolean mutable = !(dataTable.isImmutableRows());
+
+                if (upgrade) {
+                    if (compaction) {
+                        setCompactionEnabled(admin, entry, false);
+                        checkAndWaitForMajorCompaction(admin, entry);
+                    }
+                    disableTable(admin, entry);
+                    modifyIndexTable(admin, indexes);
+                    if (mutable) {
+                        modifyDataTable(admin, dataTableFullName);
+                    }
+                    enableTable(admin, entry);
+                    rebuildIndexes(entry, conf);
+                    if (compaction) {
+                        setCompactionEnabled(admin, entry, true);
+                    }
+                } else {
+                    disableTable(admin, entry);
+                    if(mutable) {
+                        modifyDataTable(admin, dataTableFullName);
+                    }
+                    modifyIndexTable(admin, indexes);
```

Comment:
Could you clean IndexNameGlobalIndexCheckerEnabledMap at this point? My PR is introducing it. It is in indexutil

Reply:
@gokceni, we should define a constant for this, if it is not defined yet. The upgrade tool should sleep this constant + 1 minutes.

Reply:
Yes, it is called GLOBAL_INDEX_CHECKER_ENABLED_MAP_EXPIRATION_MIN

Reply:
The data-tables and index-tables are disabled while their metadata is changed. Let's make sure the sleeps (if necessary) factor in the time since end of disable. Just to ensure we don't sleep for too long.

Reply:
Done!